packages/sdk/src/types/runtime/assets/assetContainer.ts

```diff
@@ -0,0 +1,198 @@
+	public get meshes() { return this.assets.filter(a => a instanceof Mesh) as Mesh[]; }
+	/** A list of all prefabs in this container */
+	public get prefabs() { return this.assets.filter(a => a instanceof Prefab) as Prefab[]; }
+	/** A list of all sounds in this container */
+	public get sounds() { return this.assets.filter(a => a instanceof Sound) as Sound[]; }
+	/** A list of all textures in this container */
+	public get textures() { return this.assets.filter(a => a instanceof Texture) as Texture[]; }
+
+	/** Create a new asset container */
+	public constructor(public context: Context) {
+		this._id = UUID();
+		context.internal.assetContainers.add(this);
+	}
+
+	/**
+	 * Generate a new material
+	 * @param name The new material's name
+	 * @param definition The initial material properties
+	 */
+	public createMaterial(name: string, definition: Partial<MaterialLike>): Material {
+		const mat = new Material(this, {
+			id: UUID(),
+			name,
+			material: resolveJsonValues(definition)
+		});
+		mat.setLoadedPromise(this.sendCreateAsset(mat));
+		return mat;
+	}
+
+	/**
+	 * Load an image file and generate a new texture asset
+	 * @param name The new texture's name
+	 * @param definition The initial texture properties. The `uri` property is required.
+	 */
+	public createTexture(name: string, definition: Partial<TextureLike>): Texture {
+		const tex = new Texture(this, {
+			id: UUID(),
+			name,
+			texture: resolveJsonValues(definition)
+		});
+		tex.setLoadedPromise(this.sendCreateAsset(tex));
+		return tex;
+	}
+
+	/**
+	 * Load an audio file and generate a new sound asset
+	 * @param name The new sound's name
+	 * @param definition The initial sound properties. The `uri` property is required.
+	 */
+	public createSound(name: string, definition: Partial<SoundLike>): Sound {
+		const sound = new Sound(this, {
+			id: UUID(),
+			name,
+			sound: resolveJsonValues(definition)
+		});
+		sound.setLoadedPromise(this.sendCreateAsset(sound));
+		return sound;
+	}
+
+	/**
+	 * preload a video stream and generate a new video stream asset
+	 * @param name The new stream's name
+	 * @param definition The initial video stream properties. The `uri` property is required.
+	 */
+	public createVideoStream(name: string, definition: Partial<VideoStreamLike>): VideoStream {
+
+		const video = new VideoStream(this, {
+			id: UUID(),
+			name,
+			videoStream: resolveJsonValues(definition)
+		});
+		video.setLoadedPromise(this.sendCreateAsset(video));
+		return video;
+	}
+
+	/**
+	 * Load the assets in a glTF file by URL, and this container with the result.
+	 * @param uri The URI to a glTF model.
+	 * @param colliderType The shape of the generated prefab collider.
+	 * @returns A promise that resolves with the list of loaded assets.
+	 */
+	public async loadGltf(uri: string, colliderType?: Payloads.CreateColliderType): Promise<Asset[]> {
+		if (!this._assets) {
+			throw new Error("Cannot load new assets into an unloaded container!");
```

Comment:
>Cannot load new assets into an unloaded container! [](start = 20, length = 50)

Why not? Seems like we should support this.

Reply:
This limitation seems artificial. What would it take to remove it?

---
In reply to: [303661613](https://github.com/microsoft/mixed-reality-extension-sdk/pull/384#discussion_r303661613) [](ancestors = 303661613)

Reply:
Because semantically you're unloading the container, not assets. Shouldn't the container itself be destroyed afterward? You're right that it's artificial, but it feels right in context. Besides, it might prevent some weird timing things if somehow the create gets processed before the unload.

---
In reply to: [303665679](https://github.com/microsoft/mixed-reality-extension-sdk/pull/384#discussion_r303665679) [](ancestors = 303665679,303661613)

Reply:
We should think of it as unloading the contents of the container, not disposing the container.

---
In reply to: [303668263](https://github.com/microsoft/mixed-reality-extension-sdk/pull/384#discussion_r303668263) [](ancestors = 303668263,303665679,303661613)

Reply:
Supporting this would allow for asset container reuse, which is a nice convenience. I think we should allow it.

---
In reply to: [303675156](https://github.com/microsoft/mixed-reality-extension-sdk/pull/384#discussion_r303675156) [](ancestors = 303675156,303668263,303665679,303661613)

Reply:
Withdrawing my objection. The asset container needs to be cleaned up, and `unload` is the best place to do it.

---
In reply to: [303677566](https://github.com/microsoft/mixed-reality-extension-sdk/pull/384#discussion_r303677566) [](ancestors = 303677566,303675156,303668263,303665679,303661613)